streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -1522,6 +1525,92 @@ private void maybeScheduleFollowupRebalance(final long encodedNextScheduledRebal
+        final Set<TaskId> activeTasks = new HashSet<>();
+        for (final KafkaStreamsAssignment assignment : assignments) {
+            for (final KafkaStreamsAssignment.AssignedTask task : assignment.assignment()) {
+                if (task.type() != KafkaStreamsAssignment.AssignedTask.Type.ACTIVE) {
+                    continue;
+                }
+
+                if (activeTasks.contains(task.id())) {
+                    return AssignmentError.ACTIVE_TASK_ASSIGNED_MULTIPLE_TIMES;
+                }
+                activeTasks.add(task.id());
+            }
+        }
+
+        // Check for ACTIVE_AND_STANDBY_TASK_ASSIGNED_TO_SAME_KAFKASTREAMS
+        for (final KafkaStreamsAssignment assignment : assignments) {
+            final Set<TaskId> activeTasksForAssignment = new HashSet<>();
+            for (final KafkaStreamsAssignment.AssignedTask task : assignment.assignment()) {
+                if (task.type() == KafkaStreamsAssignment.AssignedTask.Type.ACTIVE) {
+                    activeTasksForAssignment.add(task.id());
+                }
+            }
+
+            for (final KafkaStreamsAssignment.AssignedTask task : assignment.assignment()) {
+                if (task.type() == KafkaStreamsAssignment.AssignedTask.Type.STANDBY) {
+                    if (activeTasksForAssignment.contains(task.id())) {
+                        return AssignmentError.ACTIVE_AND_STANDBY_TASK_ASSIGNED_TO_SAME_KAFKASTREAMS;
+                    }
+                }
+            }
+        }
+
+        // Check for INVALID_STANDBY_TASK
+        final Set<TaskId> standbyTasksInOutput = new HashSet<>();
+        for (final KafkaStreamsAssignment assignment : assignments) {
+            for (final KafkaStreamsAssignment.AssignedTask task : assignment.assignment()) {
+                if (task.type() == KafkaStreamsAssignment.AssignedTask.Type.STANDBY) {
+                    standbyTasksInOutput.add(task.id());
+                }
+            }
+        }
+        for (final TaskInfo task : applicationState.allTasks()) {
+            if (!task.isStateful() && standbyTasksInOutput.contains(task.id())) {
+                return AssignmentError.INVALID_STANDBY_TASK;
+            }
+        }
+
+        // Check for MISSING_PROCESS_ID
+        final Map<ProcessId, KafkaStreamsState> clientStates = applicationState.kafkaStreamsStates(false);
+        final Set<ProcessId> clientsInOutput = assignments.stream().map(KafkaStreamsAssignment::processId)
+            .collect(Collectors.toSet());
+        for (final Map.Entry<ProcessId, KafkaStreamsState> entry : clientStates.entrySet()) {
+            final ProcessId processIdInInput = entry.getKey();
+            if (!clientsInOutput.contains(processIdInInput)) {
+                return AssignmentError.MISSING_PROCESS_ID;
+            }
+        }
+
+        // Check for UNKNOWN_PROCESS_ID
+        final Set<ProcessId> clientsInInput = clientStates.entrySet().stream().map(Map.Entry::getKey)
```

Comment:
fyi java maps have a `keySet` and `valueSet` method that lets you extract just part of the map. So this could be simplified to just `clientState.keySet()` and there's no need to build up a new map.

That said, we can just directly check `if (clientState.containsKey(processIdInOutput)` so there's really no need for the `clientsInInput` variable to be introduced at all. Or better yet, since we only need the key/processId to begin with, combine both of the above suggestions and just declare the `clientStates` right off the bat as follows to really tighten things up:

```
final Set<ProcessId> clientStates = applicationState.kafkaStreamsStates(false).keySet();
```

(Just trying to keep the number of new fields introduced to a minimum since it's complicated enough already even with just the preexisting data structures)